chainer_chemistry/links/update/mpnn_update.py

```diff
@@ -0,0 +1,130 @@
+from typing import Optional  # NOQA
+
+import chainer
+from chainer import functions
+from chainer import links
+
+import chainer_chemistry
+
+
+class MPNNUpdate(chainer.Chain):
+    r"""MPNN submodule for update part.
+
+    See: Justin Gilmer+, \
+        Neural Message Passing for Quantum Chemistry. April 2017.
+        `arXiv:1704.01212 <https://arxiv.org/abs/1704.01212>`
+
+    Args:
+        hidden_dim (int): dimension of feature vector associated to
+            each atom
+        nn (~chainer.Link):
+
+    """
+
+    def __init__(self, hidden_dim=16, nn=None):
+        # type: (int, Optional[chainer.Link]) -> None
+        super(MPNNUpdate, self).__init__()
+        with self.init_scope():
+            self.message_layer = EdgeNet(out_channels=hidden_dim, nn=nn)
+            self.update_layer = links.GRU(2 * hidden_dim, hidden_dim)
+        self.hidden_dim = hidden_dim
+        self.nn = nn
+
+    def __call__(self, h, adj):
+        # type: (chainer.Variable, chainer.Variable) -> chainer.Variable
+        # adj: (mb, edge_type, node, node)
+        mb, node, ch = h.shape
+        h = self.message_layer(h, adj)  # h: (mb, node, hidden_dim*2)
+        h = functions.reshape(h, (mb * node, self.hidden_dim * 2))
+        h = self.update_layer(h)  # h: (mb*node, hidden_dim)
+        h = functions.reshape(h, (mb, node, self.hidden_dim))
+        return h
+
+    def reset_state(self):
+        self.update_layer.reset_state()
+
+
+class EdgeNet(chainer.Chain):
+    """MPNN submodule for message part.
+
+    Edge Network expands edge vector dimension to (d x d) matrix.
+    If undirected graph, adj_in and adj_out are same.
+
+    Args:
+        out_channels (int): dimension of output feature vector
+        nn (~chainer.Link):
+
+    """
+
+    def __init__(self, out_channels, nn=None):
+        # type: (int, chainer.Link) -> None
+        super(EdgeNet, self).__init__()
+        if nn is None:
+            nn = MLP(out_channels=out_channels)
+        with self.init_scope():
+            if isinstance(nn, chainer.Link):
+                self.nn_layer_in = nn
+                self.nn_layer_out = nn
+        if not isinstance(nn, chainer.Link):
```

Comment:
What types do you expect? I feel we can raise ValueError when nn is not `chainer.Link` ?